packages/terra-functional-testing/lib/services/wdio-docker-service.js

```diff
@@ -0,0 +1,167 @@
+  async initializeSwarm() {
+    const { stdout: dockerInfo } = await exec('docker info --format "{{json .}}"');
+    const { Swarm } = JSON.parse(dockerInfo);
+
+    if (Swarm.LocalNodeState !== 'active') {
+      logger.log('Initializing docker swarm.');
+
+      await exec('docker swarm init');
+    }
+  }
+
+  /**
+   * Deploys the docker stack.
+   */
+  async deployStack() {
+    // Remove the previous stack if one exists.
+    await this.removeStack();
+
+    logger.log('Deploying docker stack.');
+
+    const composeFilePath = path.resolve(__dirname, '../docker/docker-compose.yml');
+
+    await exec(`docker stack deploy -c ${composeFilePath} wdio`);
+
+    await this.awaitNetworkReady();
+  }
+
+  /**
+   * Removes the docker stack.
+   */
+  async removeStack() {
+    const { stdout: stackInfo } = await exec('docker stack ls | grep wdio || true');
+
+    if (stackInfo) {
+      logger.log('Removing docker stack.');
+
+      await exec('docker stack rm wdio');
+
+      // Ensure the services and network have been removed.
+      await this.awaitServiceRemoval();
+      await this.awaitNetworkRemoval();
+    }
+  }
+
+  /**
+   * Waits for a command to complete successfully.
+   * @param {string} command - The shell command to run.
+   * @param {func} callback - A callback function to accept or reject the result of the command. Must return a promise.
+   * @param {number} retries - The number of times to retry the command. Defaults to 30 retries.
+   * @param {number} interval - The timeout between commands in milliseconds. Defaults to every two seconds.
+   * @returns {Promise} - A promise that resolves when the callback accepts the command response.
+   */
+  async pollCommand(command, callback, retries = RETRY_COUNT, interval = POLL_INTERVAL) {
+    return new Promise((resolve, reject) => {
+      let retryCount = 0;
+      let pollTimeout = null;
+
+      const poll = async () => {
+        if (retryCount >= retries) {
+          clearTimeout(pollTimeout);
+          pollTimeout = null;
+          reject(Error(logger.format('Timeout. Exceeded retry count.')));
+        }
+
+        try {
+          const result = await exec(command);
+
+          await callback(result).then(() => resolve());
+        } catch (error) {
+          retryCount += 1;
+          pollTimeout = setTimeout(poll, interval);
+        }
+      };
+
+      pollTimeout = setTimeout(poll, interval);
+    });
+  }
+
+  /**
+   * Ensures the docker network has been shut down.
+   */
+  async awaitNetworkRemoval() {
+    await this.pollCommand('docker network ls | grep wdio || true', (result) => (
+      new Promise((resolve, reject) => {
+        const { stdout: networkStatus } = result;
+
+        // Reject if there is an active network returned.
+        if (networkStatus) {
+          reject();
+        } else {
+          resolve();
+        }
+      })));
+  }
+
+  /**
+   * Ensures the docker services have been shut down.
+   */
+  async awaitServiceRemoval() {
```

Comment:
```suggestion
  async removeService() {
```

If we want to keep this naming, then we should update `removeStack` to be `awaitStackRemoval`

Reply:
These functions don't run the command to remove the service, they're waiting/polling the service/network to verify it has been removed. Unfortunately, I don't think `removeService` is appropriate here. Do you have naming suggestions that align more closely to the behavior of the function? I'm not a huge fan of the await prefix. waitFor seemed a little verbose, but is pretty standard for this functionality.

The `docker stack rm wdio' command run in `removeStack` is used to begin the process of removing the network. These functions poll until the network is verified to be inactive.

Reply:
`ensureServiceRemoved`

Reply:
In my opinion, `ensure` sounds like a conditional check. I think these functions should indicate they are waiting for an action to complete.  `wait`, `waitFor`, `await`, `poll`. Anyone have an opinion/vote between all the options or a nomination for another option?

Reply:
maybe `waitFor`. it's not a big deal if we don't change.

Reply:
Updated the method names to be `waitFor`:

https://github.com/cerner/terra-toolkit/pull/412/commits/5081644beac4fd9e527541810ddcd4c9c934531e